llpc/context/llpcCompiler.cpp

```diff
@@ -2271,6 +2277,46 @@ void helperThreadBuildRayTracingPipelineElf(IHelperThreadProvider *helperThreadP
   helperThreadPayload->compiler->releaseContext(context);
 }
 
+// =====================================================================================================================
+// Limited implementation of Llpc::IHelperThreadProvider to support -add-rt-helpers.
+//
+// If no deferred work helper thread providers is available when additional threads are requested via -add-rt-helpers
+// then use an instances of this class to coordinate helper threads.
+class InternalHelperThreadProvider : public Llpc::IHelperThreadProvider {
+public:
+  InternalHelperThreadProvider() {}
+  ~InternalHelperThreadProvider() {}
+
+  virtual void SetTasks(ThreadFunction *pFunction, uint32_t numTasks, void *pPayload) override {
+    m_totalInstances.exchange(numTasks);
+  }
+
+  virtual bool GetNextTask(uint32_t *pTaskIndex) override {
+    assert(pTaskIndex != nullptr);
+    *pTaskIndex = m_nextInstance.fetch_add(1);
+    return (*pTaskIndex < m_totalInstances);
+  }
+
+  virtual void TaskCompleted() override {
+    uint32_t completedInstances = m_completedInstances.fetch_add(1) + 1;
+    if (completedInstances == m_totalInstances)
+      m_event.notify_all();
+  }
+
+  virtual void WaitForTasks() override {
+    std::unique_lock<std::mutex> lock(m_lock);
+    while (m_completedInstances < m_totalInstances)
+      m_event.wait(lock);
+  }
+
+private:
+  std::atomic<uint32_t> m_nextInstance;
+  std::atomic<uint32_t> m_completedInstances;
```

Comment:
Don't we need to zero-initialize (at least) these two somewhere? I don't think `std::atomic` default constructor guarantees to value-initialize the contained field.

Reply:
I will initialize to 0.